File: kotlin/coins.kt
fun main() {

  var cnt    = 0
  val input  = List(4) {
    readLine()!!.toInt()
  }

  for (l in 0..input[0]) {
    for (m in 0..input[1]) {
      for (n in 0..input[2]) {
        if (input[3] == (500*l)+(100*m)+(50*n)) {
          cnt++
        }
      }
    }
  }
  
  println(cnt)

}

File: rust/coins.rs
use std::io;

fn main() {
  let mut input = String::new();

  io::stdin().read_line(&mut input)
    .expect("Failed to read line");     // 行の読み込みに失敗しました

  println!("Input: {}", input);     // 次のように予想しました: {}
}